### PR TITLE
core: FS: make dirfile interface accept empty object ID

### PR DESCRIPTION
**Pull request by jforissier**
A fix to allow persistent objects with an empty ID.

### CODE REVIEW

**Comment by Sadiq1811** [comment]
Reported-by: Sadiq Hussain <sadiq.muchumarri@intel.com>

File: core/tee/fs_dirfile.c [hunk 1]
```diff
@@ -198,26 +207,17 @@ TEE_Result tee_fs_dirfile_find(struct tee_fs_dirfile_dirh *dirh,
 			       const TEE_UUID *uuid, const void *oid,
 			       size_t oidlen, struct tee_fs_dirfile_fileh *dfh)
 {
-	TEE_Result res;
-	struct dirfile_entry dent;
+	TEE_Result res = TEE_SUCCESS;
+	struct dirfile_entry dent = { };
 	int n;
```
[the remark this answers is not in the document]
Comment (jforissier) [reply core/tee/fs_dirfile.c:212]: Yes

File: core/tee/fs_dirfile.c [hunk 2]
```diff
@@ -267,11 +297,13 @@ TEE_Result tee_fs_dirfile_rename(struct tee_fs_dirfile_dirh *dirh,
 	TEE_Result res;
 	struct dirfile_entry dent;
 
-	if (!oidlen || oidlen > sizeof(dent.oid))
+	if (oidlen > sizeof(dent.oid))
 		return TEE_ERROR_BAD_PARAMETERS;
 	memset(&dent, 0, sizeof(dent));
 	dent.uuid = *uuid;
 	memcpy(dent.oid, oid, oidlen);
+	if (!oidlen)
```
[the remark this answers is not in the document]
Comment (jforissier) [reply core/tee/fs_dirfile.c:305]: Ack

File: core/tee/fs_dirfile.c [hunk 3]
```diff
@@ -198,26 +207,17 @@ TEE_Result tee_fs_dirfile_find(struct tee_fs_dirfile_dirh *dirh,
...
 			return res;
 
-		/* TODO check this loop when oidlen == 0 */
-
-		if (!dent.oidlen && first_free == -1)
-			first_free = n;
+		if (is_free(&dent))
+			continue;
 		if (dent.oidlen != oidlen)
 			continue;
```
[the remark this answers is not in the document]
Comment (jforissier) [reply core/tee/fs_dirfile.c:223]: I agree, will fix

**Comment by jforissier** [comment]
Update: comments addressed and NULL check removed in `libutee`.

**Comment by etienne-lms** [comment]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits
"core: FS: make dirfile interface accept empty object ID".
and "libutee: TEE_{Create,Open}PersistentObject() should accept NULL object ID".

`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`
for commit "core: mm: vm_check_access_rights(): return success when len == 0"
since I did went though possible side effects in the implementation.

File: core/mm/vm.c [hunk 4]
```diff
@@ -1240,6 +1240,9 @@ TEE_Result vm_check_access_rights(const struct user_mode_ctx *uctx,
 	size_t addr_incr = MIN(CORE_MMU_USER_CODE_SIZE,
 			       CORE_MMU_USER_PARAM_SIZE);
 
+	if (!len)
```
Comment (jenswi-linaro) [inline core/mm/vm.c:1243]: We aren't as forgiving in `core_is_buffer_inside()` and friends.

Prior to this change a successful result from this function guaranteed that buffer indeed was supplied. If we are to change this we need to check all the call sites to see that we don't depend on the old behaviour.

**Review by jenswi-linaro — COMMENTED** [review]
Comment for the commit "core: mm: vm_check_access_rights(): return success when len == 0"

File: core/tee/fs_dirfile.c [hunk 5]
```diff
@@ -27,6 +27,16 @@ struct dirfile_entry {
 	uint32_t file_number;
 };
 
+/*
+ * An object can have an ID of size zero. This anonymous object is represented
+ * by oidlen == 0 and oid[0] != 0. When both are zero, the entry is not a valid
```
Comment (jenswi-linaro) [inline core/tee/fs_dirfile.c:32]: Why should anything but a specific value written earlier be accepted?

File: core/tee/fs_dirfile.c [hunk 6]
```diff
@@ -27,6 +27,16 @@ struct dirfile_entry {
...
+ * object.
+ */
+static inline bool is_free(struct dirfile_entry *dent)
```
Comment (jenswi-linaro) [inline core/tee/fs_dirfile.c:35]: "inline" isn't needed.

File: core/tee/fs_dirfile.c [hunk 7]
```diff
@@ -238,6 +238,36 @@ TEE_Result tee_fs_dirfile_find(struct tee_fs_dirfile_dirh *dirh,
...
+		}
+		if (res)
+			return res;
+
+		if (is_free(&dent) && first_free == -1)
+			first_free = n;
+	}
+
+	dfh->idx = n;
+	dfh->file_number = dent.file_number;
```
Comment (jenswi-linaro) [inline core/tee/fs_dirfile.c:265]: "file_number" and "hash" is always ignored by the caller and always invalid since they are from a free entry.

**Review by jenswi-linaro — COMMENTED** [review]
Comments for the commit "core: FS: make dirfile interface accept empty object ID"

**Review by jenswi-linaro — COMMENTED** [review]
For the commit "libutee: TEE_{Create,Open}PersistentObject() should accept NULL object ID"
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/tee/fs_dirfile.c [hunk 8]
```diff
@@ -27,6 +27,16 @@ struct dirfile_entry {
 	uint32_t file_number;
 };
 
+/*
+ * An object can have an ID of size zero. This anonymous object is represented
```
[the remark this answers is not in the document]
Comment (jforissier) [reply core/tee/fs_dirfile.c:31]: I see what you mean. I'll change.

File: core/tee/fs_dirfile.c [hunk 9]
```diff
@@ -267,11 +297,19 @@ TEE_Result tee_fs_dirfile_rename(struct tee_fs_dirfile_dirh *dirh,
...
 	dent.uuid = *uuid;
-	memcpy(dent.oid, oid, oidlen);
+	if (oidlen) {
+		memcpy(dent.oid, oid, oidlen);
+	} else {
+		/*
+		 * Null sized OID needs a non-null value to differentiate from
+		 * free OIDs. See is_free().
+		 */
+		dent.oid[0] = 1;
```
[the remark this answers is not in the document]
Comment (jforissier) [reply core/tee/fs_dirfile.c:311]: I will introduce a define.

File: core/tee/fs_dirfile.c [hunk 10]
```diff
@@ -238,6 +238,36 @@ TEE_Result tee_fs_dirfile_find(struct tee_fs_dirfile_dirh *dirh,
 	return TEE_SUCCESS;
 }
 
+static TEE_Result tee_fs_dirfile_new(struct tee_fs_dirfile_dirh *dirh,
```
[the remark this answers is not in the document]
Comment (jforissier) [reply core/tee/fs_dirfile.c:241]: Will do.

File: core/tee/fs_dirfile.c (hunk 6 above)
Comment (jforissier) [reply core/tee/fs_dirfile.c:35]: Right

**Review by jenswi-linaro — COMMENTED** [review]
For commit "core: storage: do not check object ID buffer when its size is zero"
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/tee/fs_dirfile.c [hunk 11]
```diff
@@ -238,33 +242,24 @@ TEE_Result tee_fs_dirfile_find(struct tee_fs_dirfile_dirh *dirh,
 	return TEE_SUCCESS;
 }
 
-static TEE_Result tee_fs_dirfile_new(struct tee_fs_dirfile_dirh *dirh,
-				     struct tee_fs_dirfile_fileh *dfh)
+static TEE_Result tee_fs_dirfile_find_empty_idx(struct tee_fs_dirfile_dirh *dh,
```
Comment (jenswi-linaro) [inline core/tee/fs_dirfile.c:245]: minor: A new static function doesn't usually need a long prefix, `find_empty_idx()` could have been enough.

**Review by jenswi-linaro — COMMENTED** [review]
For "core: FS: make dirfile interface accept empty object ID"
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by jforissier** [comment]
Squashed and rebased, thanks @jenswi-linaro.

@etienne-lms I have not kept your R-b for "core: FS: make dirfile interface accept empty object ID" since the commit has changed significantly.

**Pull request merged**